pennylane/measurements/measurements.py

```diff
@@ -129,8 +134,83 @@ class MeasurementProcess(ABC):
 
     # pylint:disable=too-many-instance-attributes
 
+    _obs_primitive: Optional["jax.core.Primitive"] = None
+    _wires_primitive: Optional["jax.core.Primitive"] = None
+    _mcm_primitive: Optional["jax.core.Primitive"] = None
+
     def __init_subclass__(cls, **_):
         register_pytree(cls, cls._flatten, cls._unflatten)
+        name = getattr(cls.return_type, "value", cls.__name__)
+        cls._wires_primitive = qml.capture.create_measurement_wires_primitive(cls, name=name)
+        cls._obs_primitive = qml.capture.create_measurement_obs_primitive(cls, name=name)
+        cls._mcm_primitive = qml.capture.create_measurement_mcm_primitive(cls, name=name)
+
+    @classmethod
+    def _primitive_bind_call(cls, obs=None, wires=None, eigvals=None, id=None, **kwargs):
+        """Called instead of ``type.__call__`` if ``qml.capture.enabled()``.
+
+        Measurements have three "modes":
+
+        1) Wires or wires + eigvals
+        2) Observable
+        3) Mid circuit measurements
+
+        Not all measurements support all three modes. For example, ``VNEntropyMP`` does not
+        allow being specified via an observable. But we handle the generic case here.
+
+        """
+        if cls._obs_primitive is None or cls._wires_primitive is None or cls._mcm_primitive is None:
+            return type.__call__(cls, obs=obs, wires=wires, eigvals=eigvals, id=id, **kwargs)
+        if obs is None:
+            wires = () if wires is None else wires
+            if eigvals is None:
+                return cls._wires_primitive.bind(*wires, **kwargs)  # wires
+            return cls._wires_primitive.bind(
+                *wires, eigvals, has_eigvals=True, **kwargs
+            )  # wires + eigvals
+
+        if isinstance(obs, Operator) or isinstance(
+            getattr(obs, "aval", None), qml.capture.AbstractOperator
+        ):
+            return cls._obs_primitive.bind(obs, **kwargs)
+        if isinstance(obs, (list, tuple)):
+            return cls._mcm_primitive.bind(*obs, **kwargs)  # iterable of mcms
+        return cls._mcm_primitive.bind(obs, **kwargs)  # single mcm
```

Comment:
So while we could reduce the first and last `return` branch to one, we don't do this here because we want to explicitly handle the different scenarios? :)

Reply:
Combining the wires and mcm primitives? I separated them as they are quite different processes in terms of what they represent.  An mcm measurement is a purely classical operation, where as a wires measurement is a quantum+classical operation.  Even if they return similarly shaped outputs, they take different kinds of inputs.


Or the two different mcm primitives calls?  something like:
```
obs = obs if isinstance(obs, (list, tuple)) else (obs, )
return cls._mcm_primitive.bind(*obs, **kwargs)
```
?

Reply:
Nevermind, this was a mildly confused question :facepalm: Thanks for trying to understand the question :sweat_smile: 

Could do the latter of what you said, but I don't think it matters too much. In particular, with the inline comments it's very clear as is.